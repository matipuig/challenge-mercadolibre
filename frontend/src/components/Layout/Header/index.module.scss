@use '~/scss/variables/colors' as colors;
@use '~/scss/variables/layout' as layout;

.header {
  background-color: colors.$primary;
  display: flex;
  flex-direction: row;
  height: layout.$headerHeight;
  max-width: 100%;
  width: 100%;

  &--dark {
    background-color: colors.$primary--dark;
  }

  .controlsContainer {
    align-items: center;
    display: flex;
    gap: 15px;
    height: 100%;
    margin: 0 auto;
    text-align: center;
    width: 80%;

    .searchTextForm {
      flex-grow: 1;
    }
  }
}
