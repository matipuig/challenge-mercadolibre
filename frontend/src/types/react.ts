/**
 * Contains the different types necessary to use React with typescript.
 */
import React, { PropsWithChildren, ReactElement as Element } from 'react';

export type ComponentChildren = React.ReactNode;
export type ReactElement = Element;
export type FunctionalComponent<P = unknown> = (props: P) => ReactElement | null;
export type FunctionalComponentWithChildren<P = unknown> = FunctionalComponent<
  PropsWithChildren<P>
>;
