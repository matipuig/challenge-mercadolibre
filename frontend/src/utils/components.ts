/**
 * Has components utilities.
 */

import { PUBLIC_CONFIG } from '~/config/public';

const { BASE_URL } = PUBLIC_CONFIG;

const getAppSubPath = (): string => {
  const url = new URL(BASE_URL);
  return url.pathname;
};

/**
 * Returns the URL for the content in the public dir.
 * NOTE: Because next doesn't like running in subpaths...
 * @param url URL of the public content.
 */
export const getURLForPublicContent = (url: string): string => {
  const appsubPath = getAppSubPath();
  if (appsubPath.endsWith('/') && url.startsWith('/')) {
    const cleanedUrl = url.replace('/', '');
    return `${appsubPath}${cleanedUrl}`;
  }
  if (!appsubPath.endsWith('/') && !url.startsWith('/')) {
    return `${appsubPath}/${url}`;
  }
  return `${appsubPath}${url}`;
};
