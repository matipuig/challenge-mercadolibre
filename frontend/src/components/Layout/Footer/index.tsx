/**
 * Contains the app header.
 */
import { ReactElement } from 'react';

import { getAuthor } from '~/state/hooks/searchResults';

import styles from './index.module.scss';

export const Footer = (): ReactElement => {
  const { name, lastname } = getAuthor();
  const authorName = `${name} ${lastname}`;
  return (
    <footer className={styles.footer}>
      <div className={styles.authorContainer}>{authorName}</div>
    </footer>
  );
};

export default Footer;
