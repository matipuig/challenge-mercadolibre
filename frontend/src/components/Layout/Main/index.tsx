/**
 * Contains the main container.
 */
import { PropsWithChildren, ReactElement } from 'react';

import styles from './index.module.scss';

export const Main = ({ children }: PropsWithChildren<unknown>): ReactElement => (
  <main className={styles.main} data-testid="main">
    <div className={styles.componentsContainer}>{children}</div>
  </main>
);

export default Main;
