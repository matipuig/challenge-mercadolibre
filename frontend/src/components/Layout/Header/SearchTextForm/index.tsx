/**
 * Contains the text search form.
 */
import { ReactElement } from 'react';

import Image from 'next/image';
import { useRouter } from 'next/router';
import { useForm } from 'react-hook-form';
import { useTranslation } from 'react-i18next';

import { CONSTANTS } from '~/constants';
import { getAvailableI18nTexts } from '~/i18n';
import { getQueryParamValue } from '~/utils/queryParams';
import { getURLForPublicContent } from '~/utils/components';

import styles from './index.module.scss';

interface SearchFormData {
  search: string;
}
const { ROUTES } = CONSTANTS;
const { QUERY_PARAMS } = ROUTES;

const getDestinationRoute = (data: SearchFormData): string => {
  const { search } = data;
  const queryParams = new URLSearchParams();
  queryParams.set(QUERY_PARAMS.SEARCH, search);
  return `${ROUTES.ITEMS}?${queryParams}`;
};

export const SearchTextForm = (): ReactElement => {
  const { t } = useTranslation();
  const texts = getAvailableI18nTexts();
  const { placeholder, tooltip } = texts.components.layout.header.searchTextForm;
  const router = useRouter();
  const search = getQueryParamValue(router.query, QUERY_PARAMS.SEARCH) || '';
  const defaultValues = { search };
  const executeSubmit = (data: SearchFormData) => {
    const destination = getDestinationRoute(data);
    router.push(destination);
  };
  const { register, handleSubmit } = useForm({ defaultValues });
  return (
    <form onSubmit={handleSubmit(executeSubmit)} className={styles.searchForm}>
      <input
        {...register('search')}
        className={styles.searchInputText}
        placeholder={t(placeholder)}
      />
      <button
        type="submit"
        onClick={handleSubmit(executeSubmit)}
        className={styles.searchButton}
        title={t(tooltip)}
      >
        <Image
          src={getURLForPublicContent('/images/icons/search.svg')}
          width={32}
          height={32}
          alt=""
          quality={100}
        />
      </button>
    </form>
  );
};

export default SearchTextForm;
