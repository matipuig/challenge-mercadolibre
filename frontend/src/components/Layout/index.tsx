/**
 * Contains the app screen layout.
 */

import { Fragment, PropsWithChildren, ReactElement } from 'react';
import Header from './Header';
import Footer from './Footer';
import Main from './Main';

export const Layout = ({ children }: PropsWithChildren): ReactElement => {
  return (
    <Fragment>
      <Header />
      <Main>{children}</Main>
      <Footer />
    </Fragment>
  );
};

export default Layout;
