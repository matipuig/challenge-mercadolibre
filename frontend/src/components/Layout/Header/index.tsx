/**
 * Contains the app header.
 */
import { ReactElement } from 'react';

import Image from 'next/image';
import Link from 'next/link';
import { useTranslation } from 'react-i18next';

import { CONSTANTS } from '~/constants';
import { getAvailableI18nTexts } from '~/i18n';
import { getURLForPublicContent } from '~/utils/components';

import styles from './index.module.scss';
import { SearchTextForm } from './SearchTextForm';

const { ROUTES } = CONSTANTS;
const texts = getAvailableI18nTexts();

export const Header = (): ReactElement => {
  const [t] = useTranslation();
  const { logo } = texts.components.layout.header;
  return (
    <header className={styles.header} data-testid="header">
      <div className={styles.controlsContainer}>
        <Link href={ROUTES.MAIN}>
          <a aria-label={t(logo.alt)}>
            <div className={styles.logo}>
              <Image
                alt={t(logo.alt)}
                aria-hidden="true"
                height={52}
                quality={100}
                src={getURLForPublicContent('/images/logos/logo-medium.png')}
                width={76}
                data-testid="logo"
              />
            </div>
          </a>
        </Link>
        <div className={styles.searchTextForm}>
          <SearchTextForm />
        </div>
      </div>
    </header>
  );
};
