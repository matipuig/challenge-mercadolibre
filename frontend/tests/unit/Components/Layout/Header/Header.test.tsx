/**
 * Contains the unit tests for the header.
 */
import { cleanup, render, RenderResult } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import { Header } from '../../../../../src/components/Layout/Header';
import { mockNextUseRouter } from '../../../../utils';

describe('Layout header', () => {
  let header: RenderResult;
  mockNextUseRouter({ query: {} });

  beforeEach(() => {
    header = render(<Header />);
  });

  afterEach(() => {
    cleanup();
  });

  it('renders.', () => {
    expect(header.baseElement).toBeInTheDocument();
  });

  it('matches snapshot.', () => {
    expect(header).toMatchSnapshot();
  });

  it('has the main logo.', () => {
    const logo = header.getByTestId('logo');
    expect(logo).toBeDefined();
  });

  it('has the link for the main page', () => {
    const link = header.getByRole('link');
    expect(link.getAttribute('href')).toEqual('/');
  });
});
