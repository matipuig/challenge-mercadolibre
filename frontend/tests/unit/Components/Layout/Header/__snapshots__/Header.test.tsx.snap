// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout header matches snapshot. 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <header
        data-testid="header"
      >
        <div>
          <a
            href="/"
          >
            <div>
              <span
                style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
              >
                <span
                  style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
                >
                  <img
                    alt=""
                    aria-hidden="true"
                    src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%2776%27%20height=%2752%27/%3e"
                    style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
                  />
                </span>
                <img
                  alt="Logo de MercadoLibre"
                  aria-hidden="true"
                  data-nimg="intrinsic"
                  data-testid="logo"
                  decoding="async"
                  src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                  style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
                />
                <noscript />
              </span>
            </div>
          </a>
          <div>
            <form
              data-testid="searchTextForm"
            >
              <input
                data-testid="originalInputValueFromQuery"
                type="hidden"
                value=""
              />
              <input
                maxlength="200"
                name="search"
                placeholder="Busc치 productos, marcas y m치s..."
              />
              <button
                title="Buscar"
                type="submit"
              >
                <span
                  style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
                >
                  <span
                    style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
                  >
                    <img
                      alt=""
                      aria-hidden="true"
                      src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%2732%27%20height=%2732%27/%3e"
                      style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
                    />
                  </span>
                  <img
                    alt=""
                    data-nimg="intrinsic"
                    decoding="async"
                    src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                    style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
                  />
                  <noscript />
                </span>
              </button>
            </form>
          </div>
        </div>
      </header>
    </div>
  </body>,
  "container": <div>
    <header
      data-testid="header"
    >
      <div>
        <a
          href="/"
        >
          <div>
            <span
              style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
            >
              <span
                style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
              >
                <img
                  alt=""
                  aria-hidden="true"
                  src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%2776%27%20height=%2752%27/%3e"
                  style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
                />
              </span>
              <img
                alt="Logo de MercadoLibre"
                aria-hidden="true"
                data-nimg="intrinsic"
                data-testid="logo"
                decoding="async"
                src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
              />
              <noscript />
            </span>
          </div>
        </a>
        <div>
          <form
            data-testid="searchTextForm"
          >
            <input
              data-testid="originalInputValueFromQuery"
              type="hidden"
              value=""
            />
            <input
              maxlength="200"
              name="search"
              placeholder="Busc치 productos, marcas y m치s..."
            />
            <button
              title="Buscar"
              type="submit"
            >
              <span
                style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
              >
                <span
                  style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
                >
                  <img
                    alt=""
                    aria-hidden="true"
                    src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%2732%27%20height=%2732%27/%3e"
                    style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
                  />
                </span>
                <img
                  alt=""
                  data-nimg="intrinsic"
                  decoding="async"
                  src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                  style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
                />
                <noscript />
              </span>
            </button>
          </form>
        </div>
      </div>
    </header>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
